source/dem/ray_tracing.cc

```diff
@@ -0,0 +1,809 @@
+                     DEMProperties::PropertiesIndex::n_properties)
+  , photon_handler(triangulation, mapping, dim)
+  , computing_timer(this->mpi_communicator,
+                    this->pcout,
+                    TimerOutput::summary,
+                    TimerOutput::wall_times)
+  , background_dh(triangulation)
+  , photon_displacement_vector(
+      parameters.ray_tracing_info.photon_displacement_vector)
+{
+  if (parameters.model_parameters.load_balance_method ==
+        Parameters::Lagrangian::ModelParameters<
+          dim>::LoadBalanceMethod::dynamic_with_sparse_contacts ||
+      parameters.model_parameters.load_balance_method ==
+        Parameters::Lagrangian::ModelParameters<
+          dim>::LoadBalanceMethod::dynamic)
+    throw std::runtime_error(
+      "The \"dynamic\" and \"dynamic_with_sparse_contacts\""
+      "load balancing methods are not supported");
+}
+
+template <int dim>
+void
+RayTracingSolver<dim>::setup_parameters()
+{
+  // Print simulation starting information
+  pcout << std::endl;
+  std::stringstream ss;
+  ss << "Running on " << n_mpi_processes << " rank(s)";
+  announce_string(pcout, ss.str(), '*');
+
+  // Check if the output directory exists
+  std::string output_dir_name = parameters.simulation_control.output_folder;
+  struct stat buffer;
+
+  // If output directory does not exist, create it
+  if (this_mpi_process == 0)
+    {
+      if (stat(output_dir_name.c_str(), &buffer) != 0)
+        {
+          create_output_folder(output_dir_name);
+        }
+    }
+
+  // Get the pointer of the only instance of the action manager
+  action_manager = DEMActionManager::get_action_manager();
+
+  // Set the simulation control as SimulationControlAdjointSteady
+  simulation_control = std::make_shared<SimulationControlSteady>(
+    parameters.simulation_control);
+
+  // Setup load balancing parameters and attach the correct functions to the
+  // signals inside the triangulation
+  // parameters.model_parameters.load_balance_method =
+  // Parameters::Lagrangian::ModelParameters<dim>::frequent;
+  load_balancing.set_parameters(parameters.model_parameters);
+
+  AdaptiveSparseContacts<dim, DEMProperties::PropertiesIndex>
+    dummy_sparse_contacts_object;
+  load_balancing.copy_references(simulation_control,
+                                 triangulation,
+                                 photon_handler,
+                                 dummy_sparse_contacts_object);
+  load_balancing.connect_weight_signals();
+}
+
+template <int dim>
+std::shared_ptr<Insertion<dim, DEMProperties::PropertiesIndex>>
+RayTracingSolver<dim>::set_particle_insertion_type()
+{
+  using namespace Parameters::Lagrangian;
+  typename InsertionInfo<dim>::InsertionMethod insertion_method =
+    parameters.particle_insertion_info.insertion_method;
+
+  std::vector<std::shared_ptr<Distribution>>
+    dummy_size_distribution_object_container;
+
+  switch (insertion_method)
+    {
+      case InsertionInfo<dim>::InsertionMethod::file:
```

Comment:
So you re-insert the particles from a concluded DEM simulation? Interesting. It would really be interesting if we eventually wrote a new vtu_insertion that would use a vtu file directly instead of forcing us to convert to our cusstom format, but this is clearly outside of the current PR. Maybe something to open an issue for.